src/flag_attn/flash.py

```diff
@@ -8,22 +8,40 @@
 
 __all__ = ["attention"]
 
+
+def maybe_contiguous(x):
+    # only when the inner most dimension is contiguous can LDGSTS be used
+    # so inner-dimension contiguity is enforced.
+    return x.contiguous() if x.stride(-1) != 1 else x
+
+def rounded_multiple(a, b):
+    return (a + b - 1) // b * b
+
 # --------------------------- public API ---------------------------
 class FlashAttention(torch.autograd.Function):
     @staticmethod
     def forward(ctx, q, k, v, causal, sm_scale, return_log_normalizer, return_total_attention):
+        # size, stride, dtype checking
         Dq, Dk, Dv = q.shape[-1], k.shape[-1], v.shape[-1]
```

Comment:
I think we could first run an all equal test over the input shapes, using iterator, as a fast path, to avoid interpreter overhead.